[BC-BREAKING] Changed tensor comparison return type from uint8 to bool

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22261 Updated docs and added deprecation warnings to acknowledge a bool tensor
* **#21113 [BC-BREAKING] Changed tensor comparison return type from uint8 to bool**

---------------
**This is a BC breaking PR**

**Description**
As part of ongoing work on adding boolean tensor to PyTorch, we are changing return type of all comparison ops from uint8 to bool.
Example: 
Old behavior: (torch.tensor([1,2,3]) > 2) --> tensor([0, 0, 1], dtype=torch.uint8)
New behavior: torch.tensor([1,2,3]) > 2) --> tensor([False, False, True], dtype=torch.bool)

Fix: 
- change comparison ops return type
- add overload for the torch.byte result which is being passed into "out=" parameter for for torch.gt/ge/lt/le/ne/eq
- added unit tests

Differential Revision: [D15552204](https://our.internmc.facebook.com/intern/diff/D15552204)

izdeby committed:
Changed tensor comparison return type from uint8 to bool

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby commented:
@pytorchbot retest this please

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

ezyang commented:
This is the big kahuna diff, right?!

ezyang commented on aten/src/ATen/native/Itertools.cpp:
```diff
@@ -51,7 +51,7 @@ Tensor combinations(const Tensor& self, int64_t r, bool with_replacement) {
   Tensor mask = _triu_mask(num_elements, r, with_replacement, self.options());
   for(Tensor &t : grids) {
-    t = t.masked_select(mask);
+    t = t.masked_select(mask.to(at::kByte));
```
Going to fix this later?

ezyang replied:
I'm going to call this Note "masked_select does not accept bool yet" and reference this at other locations I see it

ezyang replied:
Wouldn't it make sense to first do all non-BC breaking bool changes first (e.g., make `masked_select` accept bool along with byte), before making the BC-breaking bool changes (making comparison ops return bool)

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -126,7 +126,7 @@ Tensor matrix_rank(const Tensor& self, double tol, bool symmetric) {
 
   Tensor S = _matrix_rank_helper(self, symmetric);
-  return (S > tol).sum();
+  return (S > tol).to(at::kByte).sum();
```
`torch.sum` takes a desired dtype, right? Would be better to provide that instead, which saves you from having to do this conversion, and then do another promotion (I am assuming there's a promotion going on here, but I haven't checked...)

ezyang replied:
I'm going to call this Note "Sum into dtype" and just reference this in other locations I see it.

ezyang commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -136,7 +136,7 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
   Tensor S = _matrix_rank_helper(self, symmetric);
   double tol = _get_epsilon(self.scalar_type()) * std::max(self.size(0), self.size(1));
-  return (S > S.max().mul_(tol)).sum();
+  return (S > S.max().mul_(tol)).to(at::kByte).sum();
```
Note "Sum into dtype"

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -435,7 +435,7 @@ Tensor& logsumexp_out(Tensor& result, const Tensor &self, IntArrayRef dims, bool
     auto maxes = at::max_values(self, dims, true);
     auto maxes_squeezed = (keepdim ? maxes : squeeze_multiple(maxes, dims));
-    maxes_squeezed.masked_fill_(maxes_squeezed.abs() == INFINITY, 0);
+    maxes_squeezed.masked_fill_((maxes_squeezed.abs() == INFINITY).to(at::kByte), 0);
```
Note "masked functions do not accept bool yet"

ezyang commented on aten/src/ATen/test/atest.cpp:
```diff
@@ -36,7 +36,7 @@ TEST(atest, atest) {
   ASSERT_EQ(b, 4);
 
-  foo = (foo * foo) == (foo.pow(3));
+  foo = ((foo * foo) == (foo.pow(3))).to(kByte);
```
goofy but this test doesn't make sense anyway, no problems here.

ezyang commented on test/expect/TestScript.test_listconstruct_erasure.expect:
```diff
@@ -12,7 +12,7 @@ ModelProto {
         Node {type: "Less", inputs: [0,1], outputs: [2], attributes: []},
         Node {type: "Cast", inputs: [2], outputs: [3], attributes: [{ name: 'to', type: int, value: 2}]},
-        Node {type: "Cast", inputs: [3], outputs: [4], attributes: [{ name: 'to', type: int, value: 2}]},
+        Node {type: "Cast", inputs: [3], outputs: [4], attributes: [{ name: 'to', type: int, value: 9}]},
```
Do we know how this is going to affect saved JIT model backwards compatibility?

ezyang commented on test/test_distributed.py:
```diff
@@ -645,7 +645,7 @@ def _test_broadcast_helper(
                     dist.broadcast(tensor, src, group_id)
                     self.assertEqual(tensor.size(), expected_tensor.size())
-                    self.assertEqual(tensor.ne(expected_tensor).max(), 0)
+                    self.assertEqual(tensor.ne(expected_tensor).max(), torch.tensor(False, dtype=torch.bool))
```
I feel like passing `False` here should be OK. But I guess this is testing code so it doesn't matter too much.

ezyang commented on test/test_distributions.py:
```diff
@@ -2243,8 +2243,8 @@ def test_beta_underflow(self):
-            frac_zeros = float((beta_samples < 0.1).sum()) / num_samples
-            frac_ones = float((beta_samples > 0.9).sum()) / num_samples
+            frac_zeros = float((beta_samples < 0.1).byte().sum()) / num_samples
+            frac_ones = float((beta_samples > 0.9).byte().sum()) / num_samples
```
Note "Sum into dtype"

ezyang commented on test/test_distributions.py:
```diff
@@ -2254,11 +2254,11 @@ def test_beta_underflow_gpu(self):
-        self.assertEqual((beta_samples == 0).sum(), 0)
-        self.assertEqual((beta_samples == 1).sum(), 0)
+        self.assertEqual((beta_samples == 0).byte().sum(), 0)
+        self.assertEqual((beta_samples == 1).byte().sum(), 0)
```
Note "Sum into dtype"

ezyang commented on test/test_distributions.py:
```diff
@@ -2254,11 +2254,11 @@ def test_beta_underflow_gpu(self):
-        frac_zeros = float((beta_samples < 0.1).sum()) / num_samples
-        frac_ones = float((beta_samples > 0.9).sum()) / num_samples
+        frac_zeros = float((beta_samples < 0.1).byte().sum()) / num_samples
+        frac_ones = float((beta_samples > 0.9).byte().sum()) / num_samples
```
Note "Sum into dtype"

ezyang commented on test/test_indexing.py:
```diff
@@ -86,7 +86,7 @@ def test_multiple_byte_mask(self):
         v = torch.randn(5, 7, 3)
         c = torch.randn(5, 7)
-        num_ones = (c > 0).sum()
+        num_ones = (c > 0).byte().sum()
```
Note "Sum into dtype"

ezyang commented on torch/_tensor_str.py:
```diff
@@ -84,7 +84,7 @@ def __init__(self, tensor):
 
         else:
-            nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
+            nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view).byte() & tensor_view.ne(0).byte())
```
Note "masked functions do not accept bool yet"

ezyang commented on torch/csrc/jit/symbolic_script.cpp:
```diff
@@ -776,7 +776,7 @@ const std::vector<std::string> functions = {
             def backward(grad_output):
                 grad_self = (grad_output * condition.type_as(grad_output))._grad_sum_to_size(self_size)
-                grad_other = (grad_output * (1 - condition).type_as(grad_output))._grad_sum_to_size(other_size)
+                grad_other = (grad_output * (1 - condition.byte()).type_as(grad_output))._grad_sum_to_size(other_size)
```
Shouldn't we just use logical negation here?

gchanan replied:
are you tracking this somewhere in an issue?

gchanan replied:
did you answer this?

ezyang commented on torch/distributions/constraints.py:
```diff
@@ -269,9 +269,9 @@ class _LowerCholesky(Constraint):
     def check(self, value):
         value_tril = value.tril()
-        lower_triangular = (value_tril == value).view(value.shape[:-2] + (-1,)).min(-1)[0]
+        lower_triangular = (value_tril == value).byte().view(value.shape[:-2] + (-1,)).min(-1)[0]
```
A better fix is to change `min(-1)` into `any(-1)`, no?

ezyang commented on torch/distributions/constraints.py:
```diff
@@ -269,9 +269,9 @@ class _LowerCholesky(Constraint):
+        lower_triangular = (value_tril == value).byte().view(value.shape[:-2] + (-1,)).min(-1)[0]
 
-        positive_diagonal = (value.diagonal(dim1=-2, dim2=-1) > 0).min(-1)[0]
+        positive_diagonal = (value.diagonal(dim1=-2, dim2=-1) > 0).byte().min(-1)[0]
```
Ditto here

ezyang commented on torch/distributions/fishersnedecor.py:
```diff
@@ -50,13 +50,13 @@ def expand(self, batch_shape, _instance=None):
     def mean(self):
         df2 = self.df2.clone()
-        df2[df2 <= 2] = nan
+        df2[(df2 <= 2).byte()] = nan
```
Note "masked functions do not accept bool yet"

ezyang replied:
Actually, breakage of this code is very serious, and I think a very strong argument that you have to fix the masked functions first before landing this diff.

ezyang commented on torch/distributions/fishersnedecor.py:
```diff
@@ -50,13 +50,13 @@ def expand(self, batch_shape, _instance=None):
     def variance(self):
         df2 = self.df2.clone()
-        df2[df2 <= 4] = nan
+        df2[(df2 <= 4).byte()] = nan
```
Note "masked functions do not accept bool yet"

ezyang commented on torch/distributions/geometric.py:
```diff
@@ -89,7 +89,7 @@ def log_prob(self, value):
             self._validate_sample(value)
         value, probs = broadcast_all(value, self.probs.clone())
-        probs[(probs == 1) & (value == 0)] = 0
+        probs[(probs == 1).byte() & (value == 0).byte()] = 0
```
Note "masked functions do not accept bool yet"... maybe???

ezyang commented on torch/distributions/half_cauchy.py:
```diff
@@ -50,7 +50,7 @@ def variance(self):
     def log_prob(self, value):
         log_prob = self.base_dist.log_prob(value) + math.log(2)
-        log_prob[value.expand(log_prob.shape) < 0] = -inf
+        log_prob[(value.expand(log_prob.shape) < 0).byte()] = -inf
```
Note "masked functions do not accept bool yet"

ezyang commented on torch/distributions/half_normal.py:
```diff
@@ -50,7 +50,7 @@ def variance(self):
     def log_prob(self, value):
         log_prob = self.base_dist.log_prob(value) + math.log(2)
-        log_prob[value.expand(log_prob.shape) < 0] = -inf
+        log_prob[(value.expand(log_prob.shape) < 0).byte()] = -inf
```
Note "masked functions do not accept bool yet"

ezyang commented on torch/distributions/kl.py:
```diff
@@ -177,11 +177,11 @@ def kl_divergence(p, q):
-    t1[q.probs == 0] = inf
-    t1[p.probs == 0] = 0
+    t1[(q.probs == 0).byte()] = inf
+    t1[(p.probs == 0).byte()] = 0
```
Note "masked functions do not accept bool yet"

ezyang commented on torch/distributions/kl.py:
```diff
@@ -177,11 +177,11 @@ def kl_divergence(p, q):
-    t2[q.probs == 1] = inf
-    t2[p.probs == 1] = 0
+    t2[(q.probs == 1).byte()] = inf
+    t2[(p.probs == 1).byte()] = 0
```
Note "masked functions do not accept bool yet"

ezyang commented on torch/distributions/kl.py:
```diff
@@ -204,7 +204,7 @@ def _kl_binomial_binomial(p, q):
         raise NotImplementedError('KL between Binomials where q.total_count > p.total_count is not implemented')
     kl = p.total_count * (p.probs * (p.logits - q.logits) + (-p.probs).log1p() - (-q.probs).log1p())
-    inf_idxs = p.total_count > q.total_count
+    inf_idxs = (p.total_count > q.total_count).byte()
```
Note "masked functions do not accept bool yet"

ezyang commented on torch/distributions/kl.py:
```diff
@@ -411,7 +411,7 @@ def _kl_pareto_pareto(p, q):
     t2 = -alpha_ratio.log()
     result = t1 + t2 + alpha_ratio - 1
-    result[p.support.lower_bound < q.support.lower_bound] = inf
+    result[(p.support.lower_bound < q.support.lower_bound).byte()] = inf
```
Note "masked functions do not accept bool yet"

izdeby replied:
Masked functions and indexing are supported for bool masks.
One of the previous PRs in this stack fixed this issue - indexing on scalars.

ezyang commented on torch/distributions/kl.py:
```diff
@@ -481,7 +481,7 @@ def _kl_beta_normal(p, q):
 def _kl_beta_uniform(p, q):
     result = -p.entropy() + (q.high - q.low).log()
-    result[(q.low > p.support.lower_bound) | (q.high < p.support.upper_bound)] = inf
+    result[(q.low > p.support.lower_bound).byte() | (q.high < p.support.upper_bound).byte()] = inf
```
I'm going to stop labeling these, there are a lot of these sites.

ezyang commented:
If any of the BC-breaking changes are intended to be durable, I expect to see detailed information in the commit message about how to fix old code with the old behavior.

ezyang commented:
OK, in that case, are these changes necessary?

Excerpts from iurii zdebskyi's message of 2019-06-03 09:11:10 -0700:
> izdeby commented on this pull request.
> 
> > @@ -411,7 +411,7 @@ def _kl_pareto_pareto(p, q):
>      t1 = q.alpha * scale_ratio.log()
>      t2 = -alpha_ratio.log()
>      result = t1 + t2 + alpha_ratio - 1
> -    result[p.support.lower_bound < q.support.lower_bound] = inf
> +    result[(p.support.lower_bound < q.support.lower_bound).byte()] = inf
> 
> Masked functions and indexing are supported for bool masks.
> One of the previous PRs in this stack fixed this issue - indexing on scalars.
>

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

ezyang commented:
Much better! But I was under the impression that sums on uint8 promoted into int64 previously; if that's the case, `ScalarType::Int` is the wrong dtype to use.

ezyang reviewed (changes requested):
Either accumulate into int64, or add a prominent comment saying that we've changed the return precision of some operations from int64 to int32

izdeby committed:
Update on "Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

ezyang commented on test/test_torch.py:
```diff
@@ -9186,7 +9186,7 @@ def _test_bernoulli(self, t_dtype, p_dtype, device):
 
         def isBinary(t):
-            return torch.ne(t, 0).mul_(torch.ne(t, 1)).sum().item() == 0
+            return (torch.ne(t, 0) == torch.ne(t, 1)).all().item() == False
```
```
./test/test_torch.py:9189:68: E712 comparison to False should be 'if cond is False:' or 'if not cond:'
```

ezyang reviewed (approved):
please fix lint before landing

ezyang commented:
And, for the record, we're going to subsequently change `sum(booltensor)` to accumulate into `int64` by default, thus resolving the BC-breakage that is fixed a few lines here.

gchanan commented on aten/src/ATen/Declarations.cwrap:
```diff
@@ -364,6 +364,7 @@
 [[
   name: _th_equal
   cpu_bool: True
+  cuda_bool: True
```
this is orthogonal right?  I.e. can be in a separate PR, there's nothing that has to change as a result of comparison op return types changing.

izdeby replied:
https://github.com/pytorch/pytorch/pull/21538

gchanan commented on aten/src/ATen/Declarations.cwrap:
```diff
@@ -486,6 +487,8 @@
 [[
   name: _th_ixor_
   cname: __ixor__
+  cpu_bool: True
```
same here?

izdeby committed:
Update on "[BC-BREAKING] [WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[BC-BREAKING] [WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[BC-BREAKING] [WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[BC-BREAKING] [WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[BC-BREAKING] [WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

gchanan commented on tools/autograd/templates/python_variable_methods.cpp:
```diff
@@ -291,8 +291,8 @@ static Tensor dispatch_invert(const Tensor & self) {
   auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
-  if (self_.scalar_type() != at::kByte) {
-    throw TypeError("~ (operator.invert) is only implemented on byte tensors");
+  if (self_.scalar_type() != at::kByte && self_.scalar_type() != at::kBool) {
```
ok, can you file an issue?  This looks questionable because it would seem the JIT would have to implement the same logic for no good reason.

izdeby replied:
https://github.com/pytorch/pytorch/issues/22262

gchanan reviewed (changes requested):
I would expect more docs to change?  For example, we have docs for `torch.gt` that look wrong after this change.

apaszke reviewed (commented):
This also requires an update to the shape inference code in the JIT

izdeby commented:
> This also requires an update to the shape inference code in the JIT
hey, @apaszke, im not very familiar with the JIT codebase yet, can you, please, point me to a place which needs to be updated? Would you like it to be part of this PR or a separate as a prep work?

izdeby committed:
Update on "[BC-BREAKING] [WIP] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby commented:
> I would expect more docs to change? For example, we have docs for `torch.gt` that look wrong after this change.

@gchanan, here is the docs [pr](https://github.com/pytorch/pytorch/pull/22261)

apaszke commented:
@izdeby it needs to happen inside this PR, or the JIT will be broken until it is fixed. Fortunately I think the change is as simple as updating [this line](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/shape_analysis.cpp#L960) to the bool scalar type (if that's a thing? I haven't followed the bool tensor changes).

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby commented:
@apaszke, thanks for pointing this out. im surprised that nothing failed even without me changing the JIT part. is there any way i can test it besides the test_jit.py (which passes before and after the change)?

apaszke commented:
`test_jit.py` should cover that. We should open an issue about this

gchanan commented on aten/src/ATen/native/Itertools.cpp:
```diff
@@ -12,7 +12,7 @@ Tensor _triu_mask(int64_t n, int64_t dims, bool diagonal, TensorOptions opt) {
   Tensor range = at::arange(n, opt.dtype(kLong));
   std::vector<Tensor> index_grids = at::meshgrid(std::vector<Tensor>(dims, range));
-  Tensor mask = at::ones(index_grids[0].sizes(), opt.dtype(kByte));
+  Tensor mask = at::ones(index_grids[0].sizes(), opt.dtype(kBool));
```
I wonder (you don't have to make this change): does `at::full(..., True, opt.dtype(kBool));` work?  That way you avoid specifying ones for bool.

izdeby replied:
i cant replace at::ones with at::full as first will return [ Variable[CPUBoolType]{SIZE} ] and second will give [ CPUDoubleType{SIZE} ]

gchanan replied:
I think this is because you still need to pass dtype bool.

gchanan commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -136,7 +136,7 @@ Tensor matrix_rank(const Tensor& self, bool symmetric) {
   Tensor S = _matrix_rank_helper(self, symmetric);
   double tol = _get_epsilon(self.scalar_type()) * std::max(self.size(0), self.size(1));
-  return (S > S.max().mul_(tol)).sum();
+  return (S > S.max().mul_(tol)).sum(ScalarType::Long);
```
do you actually need this?  Isn't this the default now?

gchanan commented on c10/core/ScalarType.h:
```diff
@@ -109,7 +109,8 @@ namespace c10 {
   _(c10::quint8, QUInt8, i)                   \
-  _(c10::qint32, QInt32, i)
+  _(c10::qint32, QInt32, i)                   \
+  _(bool, Bool, i)
```
this doesn't seem consistent (some of the FORALLs have bool and some dont).  What makes this special?

izdeby replied:
As of right now, there are only 2 FORALL marcos that dont have Bool in it.
1) AT_FORALL_SCALAR_TYPES_EXCEPT_QINT
Its being used only in core/Scalar.h for conversion and there is a bool overload because of a compiler error. Please see a comment [there](https://github.com/pytorch/pytorch/blob/master/c10/core/Scalar.h#L38). 

2) AT_FORALL_SCALAR_TYPES_EXCEPT_HALF_AND_QINT()
I didn't touch it as there was never a real need to. its being used for in
- THBlasUtils.h 
- TensorFactories.cpp

In order to keep this PR clean, i created an [issue](https://github.com/pytorch/pytorch/issues/22744) to figure out what to do with the AT_FORALL_SCALAR_TYPES_EXCEPT_HALF_AND_QINT macro.

gchanan reviewed (approved):
I'm a bit concerned with the FORALL macros, but okay.

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

colesbury commented:
You may need to include part of https://github.com/pytorch/pytorch/pull/22326 in this PR. Basically, something like `~(tensor > 2)` should continue to work to minimize the impact of breaking backwards compatibility.

izdeby commented:
> You may need to include part of #22326 in this PR. Basically, something like `~(tensor > 2)` should continue to work to minimize the impact of breaking backwards compatibility.

Im using it this PR :) in torch/csrc/jit/symbolic_script.cpp

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

colesbury commented:
@izdeby, cool. Does it also continue to work in Python?

(In case it's not clear I had to revert https://github.com/pytorch/pytorch/pull/22326 because it changed the uint8 behavior while comparisons were still returning uint8)

ailzhang commented:
cc: @dlibenzi @asuhan This requires a followup change on XLA side once landed.

dlibenzi commented:
> cc: @dlibenzi @asuhan This requires a followup change on XLA side once landed.

Thanks @ailzhang !
I think it makes sense to give a structure to those warnings.
Like, create pytorch_xla issues referencing the pytorch PR which will be causing the issue.

dlibenzi commented:
> > cc: @dlibenzi @asuhan This requires a followup change on XLA side once landed.
> 
> Thanks @ailzhang !
> I think it makes sense to give a structure to those warnings.
> Like, create pytorch_xla issues referencing the pytorch PR which will be causing the issue.

Created https://github.com/pytorch/xla/issues/815

ailzhang commented:
@dlibenzi That's a good idea! Going forward I'm going to create issues for those that I think need XLA followup changes. 
I will only cc you/Alex for issues that I'm not certain whether it's the simple change or not. What do you think?

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

gchanan commented:
Is symbolic_variable used anywhere?

https://github.com/pytorch/pytorch/blob/9eb039334f2b84e61143363ce7f9a0aa7d6dfd2c/torch/csrc/jit/symbolic_variable.h#L81-L120 all looks wrong.

ailzhang commented:
@gchanan I think those are not in use, but good to update them for now. I had a WIP PR before aiming to remove symbolic_variable in JIT codebase, but didn't finish it due to other high pri items. I can get back to it depends on the priority.

gchanan commented:
one other note: there isn't any BC support for passing uint8 tensors to `out=` for ne, eq, etc.

colesbury commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -9,30 +9,30 @@
-    THTensor_(resizeNd)(r_, ta->dim(), THTensor_getSizePtr(ta), NULL); \
-    TH_TENSOR_APPLY3(scalar_t, r_, scalar_t, ta, scalar_t, tb, \
-                     *r__data = (*ta_data OP *tb_data) ? 1 : 0;); \
+#define TENSOR_IMPLEMENT_LOGICAL(NAME,OP)                                      \
```
I know this predates your change, but none of the instantiations of this macro are logical operators (they're relational operators). The logical operators in C are `!, &&, ||`

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Changed tensor comparison return type from uint8 to bool

gh-metadata: pytorch pytorch 21113 gh/izdeby/7/head

izdeby commented:
@pytorchbot retest this

izdeby commented:
@pytorchbot retest this please

gchanan commented on test/test_torch.py:
```diff
@@ -11366,9 +11402,20 @@ def test_bitwise_ops(self):
-        x = torch.ByteTensor([0, 1, 1])
-        self.assertEqual((~x).tolist(), [1, 0, 0])
+    def test_op_invert(self):
+        res = 0xffff - torch.arange(127, dtype=torch.int8)
```
what are you trying to test here?

gchanan replied:
I see this was just taken from https://github.com/pytorch/pytorch/pull/22320/ so you can leave it.  It does look strange though --e.g. `0xffff - torch.arange(127, dtype=torch.int8)` might not even give you an int8 once we have type promotion.

gchanan commented on aten/src/ATen/native/cuda/LegacyDefinitions.cpp:
```diff
@@ -52,4 +52,123 @@ Tensor gather_cuda(const Tensor & self, int64_t dim, const Tensor & index, bool
 
+Tensor & lt_out_cuda(Tensor & result, const Tensor & self, const Tensor & other) {
+  if (result.dtype() == at::ScalarType::Byte) {
+    AT_WARN("torch.lt received a result with dtype torch.uint8, this behavior is now deprecated," \
```
`out parameter`

izdeby commented:
@gchanan, regarding 

> did you answer this question?

 i cant replace at::ones with at::full as first will return `[ Variable[CPUBoolType]{SIZE} ]` and second will give `[ CPUDoubleType{SIZE} ]`

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21113 [WIP] Changed tensor comparison return type from uint8 to bool**
* #21096 Minor preparational JIT changes

---------------
**This is a BC breaking PR**

**Description**
As part of ongoing work on adding boolean tensor to PyTorch, we are changing return type of all comparison ops from uint8 to bool.
Example: 
Old behavior: (torch.tensor([1,2,3]) > 2) --> tensor([0, 0, 1], dtype=torch.uint8)
New behavior: torch.tensor([1,2,3]) > 2) --> tensor([False, False, True], dtype=torch.bool)

Fix: 
- change comparison ops return type
- add overload for the torch.byte result which is being passed into "out=" parameter for for torch.gt/ge/lt/le/ne/eq
- added unit tests

Differential Revision: [D15552204](https://our.internmc.facebook.com/intern/diff/D15552204)

gchanan reviewed (approved):
make the one `full` change and this looks good.

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21113 [WIP] Changed tensor comparison return type from uint8 to bool**
* #21096 Minor preparational JIT changes

---------------
**This is a BC breaking PR**

**Description**
As part of ongoing work on adding boolean tensor to PyTorch, we are changing return type of all comparison ops from uint8 to bool.
Example: 
Old behavior: (torch.tensor([1,2,3]) > 2) --> tensor([0, 0, 1], dtype=torch.uint8)
New behavior: torch.tensor([1,2,3]) > 2) --> tensor([False, False, True], dtype=torch.bool)

Fix: 
- change comparison ops return type
- add overload for the torch.byte result which is being passed into "out=" parameter for for torch.gt/ge/lt/le/ne/eq
- added unit tests

Differential Revision: [D15552204](https://our.internmc.facebook.com/intern/diff/D15552204)

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21113 [WIP] Changed tensor comparison return type from uint8 to bool**
* #21096 Minor preparational JIT changes

---------------
**This is a BC breaking PR**

**Description**
As part of ongoing work on adding boolean tensor to PyTorch, we are changing return type of all comparison ops from uint8 to bool.
Example: 
Old behavior: (torch.tensor([1,2,3]) > 2) --> tensor([0, 0, 1], dtype=torch.uint8)
New behavior: torch.tensor([1,2,3]) > 2) --> tensor([False, False, True], dtype=torch.bool)

Fix: 
- change comparison ops return type
- add overload for the torch.byte result which is being passed into "out=" parameter for for torch.gt/ge/lt/le/ne/eq
- added unit tests

Differential Revision: [D15552204](https://our.internmc.facebook.com/intern/diff/D15552204)

izdeby commented on torch/onnx/symbolic_opset8.py:
```diff
@@ -98,12 +98,12 @@ def _comparison_operator(g, input, other, op_name):
 
 # NOTE: For symbolics {gt, lt, bmm, matmul, prelu, mm, addmm, view, flatten},
 #       integer input type not supported in opset8. Cast to float if possible.
-@wrap_logical_op_with_cast_to('Byte')
```
@houseroad

BowenBao replied:
same as below

izdeby commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -769,18 +769,18 @@ def wrap_with_not(g, input, other):
     return wrap_with_not
 
 
-@wrap_logical_op_with_cast_to('Byte')
```
@houseroad

BowenBao replied:
Should be ok to remove these `@wrap_logical_op_with_cast_to('Bool')` decorators. These ONNX ops return output with dtype Bool. They were added to be compatible with PyTorch (uint8).

houseroad replied:
I think @BowenBao is right. Also I am wondering whether we should keep wrap_logical_op_with_cast_to function anymore, since it should be used anywhere if I remember correctly.

BowenBao commented on test/onnx/test_pytorch_onnx_caffe2.py:
```diff
@@ -1398,7 +1398,7 @@ def forward(self, x):
         class IsNaNModel(torch.nn.Module):
             def forward(self, input):
-                return torch.isnan(input)
+                return torch.isnan(input).byte()
```
instead of add `.byte()`, remove the line https://github.com/pytorch/pytorch/blob/b22adeb0074f57c501422a6f44c5e0ad6f1c4adb/torch/onnx/symbolic_opset9.py#L1602

houseroad replied:
do we still need this conversion if we remove the cast from symbolic?

houseroad reviewed (commented):
Aligning the logic of symbolic to PyTorch is the right action. Please make sure the test is green.

BowenBao reviewed (commented):
Added few comments on the ONNX exporter side. Also how would this change affect mask related operators? Like `masked_fill` that takes input `mask` of `ByteTensor`

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21113 [WIP] Changed tensor comparison return type from uint8 to bool**
* #21096 Minor preparational JIT changes

---------------
**This is a BC breaking PR**

**Description**
As part of ongoing work on adding boolean tensor to PyTorch, we are changing return type of all comparison ops from uint8 to bool.
Example: 
Old behavior: (torch.tensor([1,2,3]) > 2) --> tensor([0, 0, 1], dtype=torch.uint8)
New behavior: torch.tensor([1,2,3]) > 2) --> tensor([False, False, True], dtype=torch.bool)

Fix: 
- change comparison ops return type
- add overload for the torch.byte result which is being passed into "out=" parameter for for torch.gt/ge/lt/le/ne/eq
- added unit tests

Differential Revision: [D15552204](https://our.internmc.facebook.com/intern/diff/D15552204)

houseroad committed:
fix onnx expect files

houseroad committed:
fix one more onnx test

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21113 [WIP] Changed tensor comparison return type from uint8 to bool**
* #21096 Minor preparational JIT changes

---------------
**This is a BC breaking PR**

**Description**
As part of ongoing work on adding boolean tensor to PyTorch, we are changing return type of all comparison ops from uint8 to bool.
Example: 
Old behavior: (torch.tensor([1,2,3]) > 2) --> tensor([0, 0, 1], dtype=torch.uint8)
New behavior: torch.tensor([1,2,3]) > 2) --> tensor([False, False, True], dtype=torch.bool)

Fix: 
- change comparison ops return type
- add overload for the torch.byte result which is being passed into "out=" parameter for for torch.gt/ge/lt/le/ne/eq
- added unit tests

Differential Revision: [D15552204](https://our.internmc.facebook.com/intern/diff/D15552204)

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"


---------------
**This is a BC breaking PR**

**Description**
As part of ongoing work on adding boolean tensor to PyTorch, we are changing return type of all comparison ops from uint8 to bool.
Example: 
Old behavior: (torch.tensor([1,2,3]) > 2) --> tensor([0, 0, 1], dtype=torch.uint8)
New behavior: torch.tensor([1,2,3]) > 2) --> tensor([False, False, True], dtype=torch.bool)

Fix: 
- change comparison ops return type
- add overload for the torch.byte result which is being passed into "out=" parameter for for torch.gt/ge/lt/le/ne/eq
- added unit tests

Differential Revision: [D15552204](https://our.internmc.facebook.com/intern/diff/D15552204)

izdeby commented:
@pytorchbot retest this please

izdeby commented:
@ailzhang, this PR breaks XLA.

ailzhang commented:
@izdeby  Thanks! Alex had https://github.com/pytorch/xla/pull/833 ready to mitigate the breakage once this lands. :D

dlibenzi commented:
> @izdeby Thanks! Alex had [pytorch/xla#833](https://github.com/pytorch/xla/pull/833) ready to mitigate the breakage once this lands. :D

If XLA is the only broken test, you can go in, and we will follow with that PR.

izdeby committed:
Update on "[BC-BREAKING] Changed tensor comparison return type from uint8 to bool"


---------------
**This is a BC breaking PR**

**Description**
As part of ongoing work on adding boolean tensor to PyTorch, we are changing return type of all comparison ops from uint8 to bool.
Example: 
Old behavior: (torch.tensor([1,2,3]) > 2) --> tensor([0, 0, 1], dtype=torch.uint8)
New behavior: torch.tensor([1,2,3]) > 2) --> tensor([False, False, True], dtype=torch.bool)

Fix: 
- change comparison ops return type
- add overload for the torch.byte result which is being passed into "out=" parameter for for torch.gt/ge/lt/le/ne/eq
- added unit tests

Differential Revision: [D15552204](https://our.internmc.facebook.com/intern/diff/D15552204)

ezyang commented:
Lint seems to be broken by this PR
